app/code/Magento/Cron/Model/ResourceModel/Schedule/ExpressionValidator.php

```diff
@@ -0,0 +1,213 @@
+     * @SuppressWarnings(PHPMD.NPathComplexity)
+     */
+    public function checkCronExpressionPart($expr)
+    {
+        $this->lastErrorMsg = '';
+        $this->fromToModulus = [];
+
+        // handle ALL match
+        if ($this->isCronExpressionPartAsterisk($expr)) {
+            return true;
+        }
+
+        // handle multiple options
+        if ($this->isCronExpressionPartList($expr)) {
+            foreach (explode(',', $expr) as $e) {
+                if ($this->checkCronExpressionPart($e)) {
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        // handle modulus
+        $mod = 1;
+        if ($this->isCronExpressionPartModulus($expr)) {
+            $e = explode('/', $expr);
+            if (sizeof($e) !== 2) {
+                $this->lastErrorMsg = __('Invalid cron expression, expecting \'match/modulus\': %1', $expr);
+                return false;
+            }
+
+            if (!is_numeric($e[1])) {
+                $this->lastErrorMsg = __('Invalid cron expression, expecting numeric modulus: %1', $expr);
+                return false;
+            }
+
+            $expr = $e[0];
+            $mod = $e[1];
+        }
+
+        // handle all match by modulus
+        if ($this->isCronExpressionPartAsterisk($expr)) {
+            $fromValue = 0;
+            $toValue = 60;
+        } elseif ($this->isCronExpressionPartRange($expr)) {
+            // handle range
+            $e = explode('-', $expr);
+            if (sizeof($e) !== 2) {
+                $this->lastErrorMsg = __('Invalid cron expression, expecting \'from-to\' structure: %1', $expr);
+                return false;
+            }
+
+            $fromValue = $this->schedule->getNumeric($e[0]);
+            $toValue = $this->schedule->getNumeric($e[1]);
+
+            if ($fromValue > $toValue) {
+                $this->lastErrorMsg = __('Invalid cron expression, expecting from <= to in \'from-to\': %1', $expr);
+                return false;
+            }
+        } else {
+            // handle regular token
+            $fromValue = $this->schedule->getNumeric($expr);
+            $toValue = $fromValue;
+        }
+
+        if ($fromValue === false || $toValue === false) {
+            $this->lastErrorMsg = __('Invalid cron expression: %1', $expr);
+            return false;
+        }
+
+        // Store calculated from, to, modulus, to avoid duplicated logic
+        $this->fromToModulus = [$fromValue, $toValue, $mod];
+        return true;
+    }
+
+    /**
+     * @param $expr
+     * @return array
+     * @throws CronException
+     */
+    public function getFromToModulusValues($expr)
+    {
+        if (!$this->checkCronExpressionPart($expr)) {
+            throw new CronException(__($this->getLastErrorMsg()));
+        }
+
+        return $this->fromToModulus;
+    }
+
+    /**
+     * @param $expr
+     * @return bool
+     */
+    public function isCronExpressionPartAsterisk($expr)
```

Comment:
Please consider on possibility instead of exposing several public checks provide single public method that would return a *matcher* (e.g. `AsteriskMatcher`, `ListMatcher`, etc.). So code in schedule model instead of sequence of ifs would be in more OO style like:
```
return $this->expressionMatcherFactory->create($expression)->match()
```

Also, this task is out of *validator* responsibility.